sdks/java/core/src/main/java/org/apache/beam/sdk/coders/DelegateCoder.java

```diff
@@ -101,8 +104,28 @@ public Object structuralValue(T value) throws Exception {
   }
 
   @Override
+  public boolean equals(Object o) {
+    if (o == null || this.getClass() != o.getClass()) {
```

Comment:
+1 @kennknowles. Relatedly, I think most classes should either be abstract or final.

The only extender of this class is `StringDelegateCoder`, and it's only used in `Tfidf` (and clones of Tfidf). We could replace it with a proper implementation using DelegateCoder instead, or `StringDelegateCoder` could wrap it instead.